### PR TITLE
Pack ContentItem with fields

### LINKED ISSUE #12728: Reduce allocations in TokenSegment.TryMatch
[Internal Tracking issue](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1826367)

### PR DESCRIPTION
**Pull request by nkolev92**
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->

# Bug

<!-- If this is an engineering change or test change only, you do not need an issue. -->
<!-- Find or create an issue in NuGet/Home and paste the full url. -->
<!-- At the maintainers discretion, multiple changes may apply to a single issue, but only when the PRs are all created within a short period of time. -->
Fixes: https://github.com/NuGet/Home/issues/12728

## Description

The leftover allocations in the TryMatch codepaths were nearly all in the dictionaries. Given that each item in a package gets an item and all it's properties parsed, this led to many many dictionaries sized between 1-3. 

Packing helps reduce these allocations and shows some gains in time performance as well. 
To pack, we define an object property for all *known* properties and only initialize the dictionary when an item is grouped, avoiding the extra dictionaries of each item within a group.

This saves us about 40MB. 

Before:
```
Name                                                                                                                                                                                                                     	Inc %	          Inc
 nuget.packaging!NuGet.ContentModel.Infrastructure.PatternExpression+TokenSegment.TryMatch(class NuGet.ContentModel.ContentItem&,class System.String,class System.Collections.Generic.IReadOnlyDictionary`2,int32,int32&)	  0.9	   76,216,336
+ mscorlib.ni!System.Collections.Generic.Dictionary`2[System.__Canon,System.__Canon].Insert(System.__Canon, System.__Canon, Boolean)                                                                                     	  0.4	   34,115,800
+ nuget.packaging!NuGet.ContentModel.ContentItem.get_Properties()                                                                                                                                                        	  0.2	   18,992,256
+ nuget.packaging!ContentPropertyDefinition.TryLookup                                                                                                                                                                    	  0.1	   12,430,096
+ clr!JIT_New                                                                                                                                                                                                            	  0.1	   10,678,184
```

After:

```
Name                                                                                                                                                                                                                     	Inc %	          Inc
 nuget.packaging!NuGet.ContentModel.Infrastructure.PatternExpression+TokenSegment.TryMatch(class NuGet.ContentModel.ContentItem&,class System.String,class System.Collections.Generic.IReadOnlyDictionary`2,int32,int32&)	  0.5	   36,278,392
+ clr!JIT_New                                                                                                                                                                                                            	  0.3	   25,499,680
|+ clr!WKS::GCHeap::Alloc                                                                                                                                                                                                	  0.3	   25,499,680
| + clr!WKS::gc_heap::try_allocate_more_space                                                                                                                                                                            	  0.3	   25,499,680
|  + clr!GCToCLREventSink::FireGCAllocationTick_V3                                                                                                                                                                       	  0.3	   25,499,680
|   + clr!CoTemplate_qqhxpzqp                                                                                                                                                                                            	  0.3	   25,499,680
|    + clr!EtwCallout                                                                                                                                                                                                    	  0.3	   25,499,680
|     + clr!ETW::SamplingLog::SendStackTrace                                                                                                                                                                             	  0.3	   25,499,680
|      + Type NuGet.ContentModel.ContentItem                                                                                                                                                                             	  0.3	   25,499,680
+ nuget.packaging!ContentPropertyDefinition.TryLookup                                                                                                                                                                    	  0.1	   10,778,712

```

Some micro benchmarks:

Before:

| Method                | Mean          | Error      | StdDev     | Gen0       | Gen1      | Allocated    |
|---------------------- |--------------:|-----------:|-----------:|-----------:|----------:|-------------:|
| NJ                    |      40.24 us |   0.353 us |   0.331 us |     3.2959 |    0.1221 |      56.5 KB |
| MicrosoftBuildRuntime |      62.62 us |   0.822 us |   0.686 us |     8.1787 |    0.6104 |    138.23 KB |
| AllPackages           | 152,979.21 us | 804.331 us | 713.019 us | 14250.0000 | 2000.0000 | 240733.08 KB |

After:

| Method                | Mean          | Error        | StdDev       | Gen0       | Gen1      | Allocated    |
|---------------------- |--------------:|-----------:|-----------:|-----------:|----------:|-------------:|
| NJ                    |      34.69 us |   0.257 us |   0.228 us |     2.3193 |    0.0610 |     39.92 KB |
| MicrosoftBuildRuntime |      58.17 us |   0.680 us |   0.636 us |     7.3853 |    0.4883 |    125.22 KB |
| AllPackages           | 133,135.56 us | 669.977 us | 626.697 us | 11000.0000 | 1000.0000 | 185389.49 KB |

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [x] Added tests
- [x] ~Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.~

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 1]
```diff
@@ -614,7 +614,7 @@ private static void PopulatePackageFolders(IEnumerable<string> packageFolders, L
             lockFile.PackageFolders.AddRange(packageFolders.Select(path => new LockFileItem(path)));
         }
 
-        private static LockFileLibrary CreateLockFileLibrary(LocalPackageInfo package, string sha512, string path)
+        internal static LockFileLibrary CreateLockFileLibrary(LocalPackageInfo package, string sha512, string path)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs:617]: I've used this method in a lot of my profiling. I've found this extremely useful, since I can have tests and other assemblies (such as benchmarks call this method).

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs [hunk 2]
```diff
@@ -7,15 +7,142 @@ namespace NuGet.ContentModel
...
+        // These must match ManagedCodeConventions.PropertyNames. They cannot be used since they're static and switch requires const.
+        internal const string TargetFrameworkMoniker = "tfm";
+        internal const string RuntimeIdentifier = "rid";
+        internal const string AnyValue = "any";
+        internal const string ManagedAssembly = "assembly";
+        internal const string Locale = "locale";
+        internal const string MSBuild = "msbuild";
+        internal const string SatelliteAssembly = "satelliteAssembly";
+        internal const string CodeLanguage = "codeLanguage";
+
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs:21]: The empty space here separates the properties that match ManagedCodeConventions.PropertyNames and the ones that are more of an implementation detail rather than folder structure contract.

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs [hunk 3]
```diff
@@ -24,6 +151,46 @@ public override string ToString()
         private Dictionary<string, object> CreateDictionary()
         {
             var properties = new Dictionary<string, object>();
+            if (_assembly != null) // We always initialize the dictionary with the packed values.
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs:156]: The copying here preserves backwards compatibility in the APIs.

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs (hunk 2 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs:21]:
```suggestion

        // These do not map to ManagedCodeConventions, they're used internally
```

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs [hunk 4]
```diff
@@ -7,15 +7,142 @@ namespace NuGet.ContentModel
...
+
+        internal void Add(string key, object value)
+        {
+            if (_properties != null)
+            {
+                _properties.Add(key, value);
+            }
+            else
+            {
+
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs]: nit: probably unnecessary whitespace

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs:1]: This PR modifies a significant percentage of the total file.  Therefore, consider adding `#nullable enable`

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs:1]: Most of this file is new. Consider using `#nullable enable`.

**Review by zivkan — DISMISSED** [review]

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs:1]: Good idea.

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs:1]: I'll give it a shot. I don't know if I know enough to annotate this properly, but won't know if I don't try.

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItem.cs]: Most definitely unnecessary :D

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs [hunk 5]
```diff
@@ -358,69 +391,155 @@ private static bool IsValidAsset(string path)
...
                 {
-                    if (property.Key.Equals("tfm_raw", StringComparison.Ordinal))
+                    foreach (var property in obj._properties)
                     {
-                        // We store the raw version of the TFM, but we don't want it to affect the result.
-                        continue;
-                    }
 #if NETFRAMEWORK || NETSTANDARD
-                    hashCode ^= property.Key.GetHashCode();
+                        hashCode ^= property.Key.GetHashCode();
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs:406]: I know nothing about good hash code design. But our own `HashCodeCombiner`, and also [.NET's HashCode](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/HashCode.cs) both do multiplcations and bitshifts.

Should we port this to use HashCodeCombiner?

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs:406]: We could potentially yeah.

In the interest of getting this PR merged, I can do that in a follow-up.
